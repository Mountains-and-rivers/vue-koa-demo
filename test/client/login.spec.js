import Vue from 'vue'
import VueRouter from 'vue-router'
import ElementUI from 'element-ui'
import { mount } from '@vue/test-utils'
import Login from '@/components/Login'

Vue.use(ElementUI)
Vue.use(VueRouter)
const router = new VueRouter()

jest.mock('axios', () => ({
  post: jest.fn()
    // for test3
    .mockImplementationOnce(() => Promise.reject(new Error()))
    // for test4
    .mockImplementationOnce(() => Promise.resolve({
      data: {
        success: true
      }
    }))
    .mockImplementationOnce(() => Promise.resolve({
      data: {
        success: false
      }
    }))
    // for test5
    .mockImplementationOnce(() => Promise.resolve({
      data: {
        success: true
      }
    }))
    .mockImplementationOnce(() => Promise.resolve({
      data: {
        success: true
      }
    }))
}))

describe('Login.vue', () => {
  let wrapper
  beforeEach(() => {
    jest.useFakeTimers()
    wrapper = mount(Login, { sync: false, router })
  })

  test('login method should be called when clicking login button', async () => {
    const stub = jest.fn()
    wrapper.setMethods({ login: stub })
    await wrapper.vm.$nextTick()
    wrapper.find('.btns .el-button:nth-child(2)').trigger('click')
    expect(stub).toBeCalledTimes(1)
  })

  test('login failed when username and password are empty', async () => {
    const result = await wrapper.vm.login()
    expect(result).toBe(false)
  })

  test('login failed when username exists', async () => {
    wrapper.setData({
      form: {
        name: 'breezymelon',
        pwd: '111'
      }
    })
    const result = await wrapper.vm.login()
    expect(result).toBe(false)
  })

  test('login failed when password is wrong', async () => {
    wrapper.setData({
      form: {
        name: 'breezymelon',
        pwd: '121'
      }
    })
    const result = await wrapper.vm.login()
    expect(result).toBe(false)
  })

  test('login success when username and password is correct', async () => {
    wrapper.setData({
      form: {
        name: 'breezymelon',
        pwd: '111'
      }
    })
    const result = await wrapper.vm.login()
    jest.runAllTimers()
    expect(result).toBe(true)
  })

  test('login.vue snapshot test', () => {
    expect(wrapper.element).toMatchSnapshot()
  })
})
