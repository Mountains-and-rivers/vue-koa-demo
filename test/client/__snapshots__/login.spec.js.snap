// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login.vue login.vue snapshot test 1`] = `
<div
  class="login"
>
  <form
    class="el-form"
  >
    <div
      class="el-form-item el-form-item--feedback"
    >
      <label
        class="el-form-item__label"
        for="name"
        style="width: 70px;"
      >
        用户名
      </label>
      <div
        class="el-form-item__content"
        style="margin-left: 70px;"
      >
        <div
          class="el-input"
        >
          <!---->
          <input
            autocomplete="off"
            class="el-input__inner"
            type="text"
          />
          <!---->
          <!---->
          <!---->
        </div>
        <!---->
      </div>
    </div>
     
    <div
      class="el-form-item el-form-item--feedback"
    >
      <label
        class="el-form-item__label"
        for="pwd"
        style="width: 70px;"
      >
        密 码
      </label>
      <div
        class="el-form-item__content"
        style="margin-left: 70px;"
      >
        <div
          class="el-input"
        >
          <!---->
          <input
            autocomplete="off"
            class="el-input__inner"
            type="password"
          />
          <!---->
          <!---->
          <!---->
        </div>
        <!---->
      </div>
    </div>
     
    <div
      class="btns"
    >
      <button
        class="el-button el-button--default"
        type="button"
      >
        <!---->
        <!---->
        <span>
          注 册
        </span>
      </button>
       
      <button
        class="el-button el-button--primary"
        type="button"
      >
        <!---->
        <!---->
        <span>
          登 录
        </span>
      </button>
    </div>
  </form>
</div>
`;
